extras/sht3x/sht3x.c

```diff
@@ -0,0 +1,401 @@
+ * modification, are permitted provided that the following conditions are met:
+ * 
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its 
+ * contributors may be used to endorse or promote products derived from this
+ * software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * Driver for Sensirion SHT3x digital temperature and humity sensor
+ * connected to I2C
+ *
+ * Part of esp-open-rtos
+ */
+
+#include <string.h>
+
+#include "sht3x.h"
+
+#include "FreeRTOS.h"
+#include "task.h"
+
+#include "espressif/esp_common.h"
+#include "espressif/sdk_private.h"
+
+#define SHT3x_STATUS_CMD               0xF32D
+#define SHT3x_CLEAR_STATUS_CMD         0x3041
+#define SHT3x_RESET_CMD                0x30A2
+#define SHT3x_FETCH_DATA_CMD           0xE000
+#define SHT3x_HEATER_OFF_CMD           0x3066
+
+const uint16_t SHT3x_MEASURE_CMD[6][3] = { {0x2c06,0x2c0d,0x2c10},    // [SINGLE_SHOT][H,M,L]
+                                           {0x2032,0x2024,0x202f},    // [PERIODIC_05][H,M,L]
+                                           {0x2130,0x2126,0x212d},    // [PERIODIC_1 ][H,M,L]
+                                           {0x2236,0x2220,0x222b},    // [PERIODIC_2 ][H,M,L]
+                                           {0x2234,0x2322,0x2329},    // [PERIODIC_4 ][H,M,L]
+                                           {0x2737,0x2721,0x272a} };  // [PERIODIC_10][H,M,L]
+
+// tick counts [High, Medium, Low]
+// minimum is one tick if portTICK_PERIOD_MS is greater than 20 or 30 ms
+const int32_t SHT3x_MEASURE_DURATION[3] = { (30 + portTICK_PERIOD_MS-1)/portTICK_PERIOD_MS,
+                                            (20 + portTICK_PERIOD_MS-1)/portTICK_PERIOD_MS,
+                                            (20 + portTICK_PERIOD_MS-1)/portTICK_PERIOD_MS };
+
```

Comment:
Sorry to be dithering on this. Trying to work it out for myself. Tried `vTaskDelay(1)` and on a small set get times from 1531us to 9014us, and `vTaskDelay(2)` get times from 11873us to 19343us.

The datasheet has the timings 4ms, 6ms, and 15ms. Might it be best to actually use those values in the code. Looking at the test results about for a 2 tick delay it is clear that this is not adequate to ensure a 15ms delay, so it is not a surprise that was failing.

The first tick must just run to the next tick, which might be much shorter than a full tick, and in this case it should probably be assumed it could be zero time, so add one tick period for that (not less one milli-second).

Also in rare cases a tick period might be halved, as the clock might run at twice the expected speed due to a cpu freq change, so to be sure add half a tick period extra.

```
#define TIME_TO_TICKS(ms) (((ms) + portTICK_PERIOD_MS + portTICK_PERIOD_MS / 2) / portTICK_PERIOD_MS)
```

Could you use a macro to avoid duplication, and the times from the datasheet:

```
const int32_t SHT3x_MEASURE_DURATION[3] = { TIME_TO_TICKS(15), TIME_TO_TICKS(6), TIME_TO_TICKS(4) };
```

Reply:
Thank you for all your effort. Tried your formula and observed the following. For high repeatability (15 ms duration), I got the following results:

Duration=3 ticks
mean(vTaskDelay(3)) = 28.145160
min(vTaskDelay(3)) = 21.272000

Seems not to be a problem.

But for low repeatability (4 ms duration), I got:

Duration = 1 tick      (ok, according to your fomula it results from 19 ms / 10 ms)
mean(vTaskDelay) = 8.214056
min(vTaskDelay) = 2.735000 :-(

To be sure, that waiting time is long enough, evene if the shortest possible tick happens, the tick count should be 2 for low and medium repeatability and 3 for high repeatability.

Reply:
Sorry, try again.

```
#define TIME_TO_TICKS(ms) (1 + ((ms) + portTICK_PERIOD_MS / 2 + (portTICK_PERIOD_MS - 1)) / portTICK_PERIOD_MS)
```

This gives 2 ticks for 4ms, 3 ticks for 6ms, and 3 ticks for 15ms. The extra tick might be needed for 6ms because it might be unlucky and one of the ticks might be halved to 5ms, so at the extreme it might wait only 5ms. What do you think?

Reply:
I will try :-)

Reply:
What about the following approach? 

Function **_sht3x_start_measurement_** doesn't return the number of of ticks required to wait for the results. Instead, the user task just calls a new function for waiting, e.g., **_sht3x_wait_for_data_** like following:

```
sht3x_start_measurement (sensor, single_shot);
sht3x_wait_for_data (sensor);        
sht3x_get_results (sensor, &values);
```

Function **_sht3x_wait_for_data_** clould deal with ticks that were too short and could look like following or similar:
```
int32_t SHT3x_MEASURE_DURATION[3] = { 15, 6, 4 };
...
bool sht3x_wait_for_data (sht3x_sensor_t* dev)
{
    ... // some error handling

    vTaskDelay((SHT3x_MEASURE_DURATION[dev->repeatability] + portTICK_PERIOD_MS-1) / portTICK_PERIOD_MS);

    while (sdk_system_get_time()/1000 - dev->meas_start_time < SHT3x_MEASURE_DURATION[dev->repeatability])
        vTaskDelay (1);
    ...
}
```
The lookup table contains the values in ms. Function **_sht3x_wait_for_data_** computes the tick number required and uses **_vTaskDelay_** for waiting. If the ticks were too short, it simply waits one further tick.

Reply:
@ourairquality
What is your opinion on the above approach of using a specialized waiting function?

Another possible approach would be to return the ticks by the **sht3x_start_measurement** function, but to give **sht3x_get_results** an additional boolean parameter which allows the user to tell the function to wait for the results. With that approach the user would have the choice either to realize waiting by himself or to invoke **sht3x_get_results** immediately after **sht3x_start_measurement** and to let the function **sht3x_get_results** wait:
```
sht3x_start_measurement (sensor, single_shot);
sht3x_get_results (sensor, &values, true);   // third parameter tells the function to wait for the results
```
If function **sht3x_get_results** should not wait and results are not available, it fails.

Reply:
The time to complete a measurement is a constant based on the configuration so it does seem better to not return it from `sht3x_start_measurement` every time. There seem no real need for `sht3x_start_measurement` to handle any already running measurement, that seems a coding error.

Some callers might be able to pre-compute a schedule taking advantage of the constant measurement period so exposing this seems better in general. Some callers might still want to know the measurement time and to call back themselves so perhaps add a function to return the measurement period that is only valid after initialization.

Most callers probably do not care if they wait 3 ticks ot 6 ticks. Might be over optimizing here. I would just bump up the number of ticks to be always safe rather than defining a function to wait more accurately. If we want to add such a wait function then it probably has more general utility.

I would remove `sht3x_is_measuring`. It seems too overloaded. The caller should manage this.

I would remove `sht3x_get_raw_data`. It also seems too overload, handling single shot and periodic. In single shot mode it is trivial for the caller to call `sht3x_read_data` and then call a function to check the crc. For the periodic mode perhaps add back a function to fetch and read the data.

I would remove `sht3x_get_results`. It also seems too overload, and it does not help when the raw values are needed. I would not overload it even more by having it optionally wait.

If after the above clean up it seems useful to add back some higher level functions that combine these then that might be useful but I would keep it minimal.

These are just suggestions. I think the code as-is could be merged at least as a starting point.

Reply:
@ourairquality 
Here are the results of tests with your last suggested ms2ticks macro

repeatability | ticks | required | min | mean | max | read failures
-------------- | ------ | -----  | ----- | ------- | ------ | ----------------------
low | 2 | 4 ms | 13.65 ms | 18.99 ms | 23.86 ms | 0
medium | 3 | 6 ms | 24.35 ms | 28.99 ms | 33.97 ms | 0
high | 3 | 15 ms | 24.35 ms | 28.99 ms | 33.97 ms | 0

3 ticks are not necessary for medium repeatability. I think, the following macro should do the best
```
#define TIME_TO_TICKS(ms) (1 + ((ms) + (portTICK_PERIOD_MS - 1)) / portTICK_PERIOD_MS)
```
It maps the ms value to the next value divisible by portTICK_PERIOD_MS and adds one additional tick for security. That is, one tick could be 0 ms in worst case. It results in 2 ticks for 4 ms an 6 ms (which seems to be enough) and 3 ticks for 15 ms. The tests results were:

repeatability | ticks | required | min | mean | max | read failures
-------------- | ------ | -----  | ----- | ------- | ------ | ----------------------
low | 2 | 4 ms | 13.65 ms | 18.99 ms | 23.86 ms | 0
medium | 2 | 6 ms | 13.65 ms | 18.99 ms | 23.86 ms | 0
high | 3 | 15 ms | 24.35 ms | 28.99 ms | 33.97 ms | 0

Reply:
Have a look at `sdk__xt_timer_int()` in `open_esplibs/libmain/os_cpu_a.c`. This is the rtos tick handler. Note it adjusts the next trigger count based on the current cpu freq. If the cpu freq changes soon after then the tick period can be longer or shorter than expected. These cpu freq changes are not frequent, perhaps just when computing some hashes etc. This was why I suggested accounting for one tick being halved in period, so the 'medium' with only two ticks might be `0 + 5` ms and thus less than 6. Perhaps this is all being over optimized, does any app really care if it waits an extra tick to be safe?

Reply:
@ourairquality 
Many thanks for your comprehensive review and your suggestions. I will think about some changes.

Concerning the ticks you are probably right that it doesn't matter for most applications whether they wait 2 or 3 ticks. Just for the sake of interest, is it really possible that a tick will take only 0 ms? I would expect that even if the CPU frequency increases, the previously calculated tick trigger count should not lead to a tick of 0 ms.

Reply:
@ourairquality 
I am really very greatful for your suggestions and like to discuss them with you.

First I would like to explain a bit why you probably feel that some of the functions are overloaded. Most of your points result from the fact that I have learned all the time
- always realize interfaces that make them as easy as possible to use,
- always hide any details and just provide the user only with the results he is interested on
- always program fault tolerant, and
- never make any assumption that the user would be careful.

That is probably the reason, why I like abstraction and fault tolerance which leads of course to some overloading. I have played a bit with the sensor to see how fault tolerant it is against wrong timing. 

On function ```sht3x_start_measurement```: It is possible to remove the check whether a measurement is already started. The sensor doesn't react on multiple start measurement commands. Concerning the duration in ticks as return value, your are principally right. Even though it is a constant, I would still leave it as it is since the user task can use it but it isn't a must. For that users that don't want to implement a fixed schedule and need some timing information it is very useful.

On function ```sht3x_is_measuring```: I agree, it can be removed. In that case the user would have always to user ```vTaskDelay``` to wait which shouldn't be a problem. However, I would like to keep it for internal use since functions like ```sht3x_get_raw_data``` or ```sht3_get_results``` the needs to check whether measurement is still running. The trial to read out sensor data before they are available lead to an i2c communication error and the sensor blocks completely until power down. Unfortunately, the sensor has no possibility to check whether data are ready to read other than clock stretching that does not work correctly in current i2c version. Thus, this check has to be realized by timers.

On ```sht3x_get_raw_data```: I would like to leave this function as is. Reading the data and checking the CRC is identical for single shot and periodic mode. The only difference is that in periodic mode a fetching command has to be sent to the sensor before data are read. But, this is a technical detail that should be hidden by the function. Therefore, having only one function for both modes that provide exactly the same results makes sense from my point of view. Furthermore, I would like to remove the CRC data from the raw data provided to the user. I don't see any reason why the user task should need the CRCs. These are only to secure the communication on the i2c interface.

On ```sht3x_get_results```: This function is the higher level function you ask for. It combines reading the raw data and converting them into real values. Or do you mean that there should be one single function that comprises starting the measurement, waiting, reading raw data, and converting them to real values?

Reply:
@gschorcht We obviously see things differently. I don't think that is a problem. We would probably write different drivers that would appeal to different people. These are all 'extras' and there is room for different approaches. So I would encourage you to do it as you see it meeting some use cases, and do it true to that, and well.

Regarding the near zero period tick, this occurs because the time from the start of the delay to the next tick defines the first tick period and the next tick might occur just after requesting a delay in which case that first tick period becomes almost zero.

I might have even defined the wait times as run time constants.